### ISSUE TITLE
pubsub: GCP IAM resource conditions

### ISSUE THREAD

**Issue by inigohu** [googleapis/google-cloud-go#4232]
**Client**

PubSub

**Environment**

Alpine Docker on GKE

**Go Environment**

$ go version
go version go1.16.5

$ go env
GO111MODULE="on"

**Code**

```go
package main

import (
	"context"
	"fmt"
	"log"

	"cloud.google.com/go/pubsub"
)

func main() {
	ctx := context.Background()

	client, err := pubsub.NewClient(ctx, "myproject")
	if err != nil {
		log.Fatal(err)
	}
	defer client.Close()

	// Publish "hello world" on topic1.
	topic := client.Topic("test")
	res := topic.Publish(ctx, &pubsub.Message{
		Data: []byte("hello world"),
	})

	// The publish happens asynchronously.
	// Later, you can get the result from res:
	_, err = res.Get(ctx)
	if err != nil {
		log.Fatal(err)
	}

	// Use a callback to receive messages via subscription1.
	sub := client.Subscription("test")
	err = sub.Receive(ctx, func(ctx context.Context, m *pubsub.Message) {
		fmt.Println(m.Data)
		m.Ack() // Acknowledge that we've consumed the message.
	})
	if err != nil {
		log.Fatal(err)
	}
}
```

**Expected behavior**

Send and receive a message to pubsub topic/subscriber.

**Actual behavior**

`rpc error: code = PermissionDenied desc = User not authorized to perform this action.`

**Screenshots**
![1](https://user-images.githubusercontent.com/6482124/121214793-bf405c00-c87f-11eb-85af-baf36619304b.jpg)
![2](https://user-images.githubusercontent.com/6482124/121214806-c2d3e300-c87f-11eb-9a2d-4a92d1d081ea.jpg)
![3](https://user-images.githubusercontent.com/6482124/121214818-c5363d00-c87f-11eb-925e-d2600c84bd73.jpg)

**Additional context**

- I created a topic and a subscription called `test` in my GCP project called `myproject`.
- I created a GCP service account and binded to my Kubernetes service account using Workload Identity.
- I assigned two roles to my GCP service account using conditions:
```
role: Pub/Sub Publisher
condition: resource.name == "projects/myproject/topics/test"
```
and
```
role: Pub/Sub Subscriber
condition: `resource.name == "projects/myproject/subscriptions/test"`
```

The assignments work because I see the service account within the topic/subscription permissions.
If I remove those conditions, my code works.

**Comment by hongalex** [comment]
Just to confirm, is the error coming from the publish or subscribe call?

**Comment by inigohu** [comment]
>Just to confirm, is the error coming from the publish or subscribe call?

Both. The message does not publish to the topic and I get the error when I try to get the result here `_, err = res.Get(ctx)` .

I also commented out the first part of my test code to receive only subscription messages, and again, I get the error here `err = sub.Receive(ctx, func(ctx context.Context, m *pubsub.Message)`.

**Comment by hongalex** [comment]
Ah, I found the problem. Pub/Sub supports resource level permissions, but not IAM resource conditions. I believe these accomplish the same thing, but are slightly different. The services that support conditions that you were using can be found [here](https://cloud.google.com/iam/docs/conditions-resource-attributes). I believe the way to add resource level permissions in Pub/Sub in the dashboardwould be to click the "Add Member" button on your last two screenshots.

We are trying to move our services over to IAM resource conditions but for now this is the workaround (Pub/Sub Lite uses IAM resource conditions as it is a newer service). Closing out this issue now since this isn't a client library issue.
